Cython/Compiler/ExprNodes.py

```diff
@@ -11377,10 +11377,22 @@ def py_operation_function(self, code):
                 is_unicode_concat = False
 
             if is_unicode_concat:
-                if self.operand1.may_be_none() or self.operand2.may_be_none():
-                    return '__Pyx_PyUnicode_ConcatSafe'
-                else:
-                    return '__Pyx_PyUnicode_Concat'
+                if self.inplace or self.operand1.is_temp:
+                    code.globalstate.use_utility_code(
+                        UtilityCode.load_cached("UnicodeConcatInPlace", "ObjectHandling.c"))
+                func = '__Pyx_PyUnicode_Concat'
+        elif type1 is bytes_type and type2 is bytes_type:
+            code.globalstate.use_utility_code(
+                    UtilityCode.load_cached("BytesConcatInPlace", "ObjectHandling.c"))
+            func = '__Pyx_PyBytes_Concat'
+
+        if func:
+            # any necessary utility code will be got by "NumberAdd" in generate_evaluation_code
+            if self.inplace or self.operand1.is_temp:
+                func += 'InPlace'  # upper case to indicate unintuitive macro
+            if self.operand1.may_be_none() or self.operand2.may_be_none():
+                func += 'Safe'
```

Comment:
I don't see a `__Pyx_PyUnicode_ConcatSafe`, but that name can result from this.

Reply:
https://github.com/cython/cython/blob/4dd7c18de76cb5c90a5168fba35990600307e642/Cython/Utility/ModuleSetupCode.c#L713 Already exists in ModuleSetupCode.py